[BUG] Databricks 11.3 - Task failed while writing rows for Delta table -  java.lang.Integer cannot be cast to java.lang.Long

**Describe the bug**
TEsting customer query on Databricks 11.3 that writes to delta table and tasks failed with:

```
org.apache.spark.SparkException: Task failed while writing rows.
        at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:349)
        at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$15(GpuFileFormatWriter.scala:262)
        at org.apache.spark.scheduler.ResultTask.$anonfun$runTask$3(ResultTask.scala:75)
        at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
        at org.apache.spark.scheduler.ResultTask.$anonfun$runTask$1(ResultTask.scala:75)
        at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:55)
        at org.apache.spark.scheduler.Task.doRunTask(Task.scala:169)
        at org.apache.spark.scheduler.Task.$anonfun$run$4(Task.scala:137)
        at com.databricks.unity.UCSEphemeralState$Handle.runWith(UCSEphemeralState.scala:41)
        at com.databricks.unity.HandleImpl.runWith(UCSHandle.scala:99)
        at com.databricks.unity.HandleImpl.$anonfun$runWithAndClose$1(UCSHandle.scala:104)
        at scala.util.Using$.resource(Using.scala:269)
        at com.databricks.unity.HandleImpl.runWithAndClose(UCSHandle.scala:103)
        at org.apache.spark.scheduler.Task.$anonfun$run$1(Task.scala:137)
        at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
        at org.apache.spark.scheduler.Task.run(Task.scala:96)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$13(Executor.scala:902)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1697)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:905)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:760)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Long
        at scala.runtime.BoxesRunTime.unboxToLong(BoxesRunTime.java:107)
        at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getWriter$5(InternalRow.scala:176)
        at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getWriter$5$adapted(InternalRow.scala:176)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.writeValueToRow(GpuStatisticsCollection.scala:270)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.computeMin(GpuStatisticsCollection.scala:313)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.$anonfun$batchStatsToRow$1(GpuStatisticsCollection.scala:215)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.$anonfun$batchStatsToRow$1$adapted(GpuStatisticsCollection.scala:215)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.$anonfun$structStatsToRow$4(GpuStatisticsCollection.scala:258)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.$anonfun$structStatsToRow$4$adapted(GpuStatisticsCollection.scala:257)
        at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
        at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
        at com.nvidia.spark.RebaseHelper$.withResource(RebaseHelper.scala:26)
       at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.$anonfun$structStatsToRow$2(GpuStatisticsCollection.scala:257)
        at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.structStatsToRow(GpuStatisticsCollection.scala:250)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.$anonfun$batchStatsToRow$5(GpuStatisticsCollection.scala:237)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.$anonfun$batchStatsToRow$5$adapted(GpuStatisticsCollection.scala:236)
        at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
        at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
        at com.nvidia.spark.RebaseHelper$.withResource(RebaseHelper.scala:26)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.batchStatsToRow(GpuStatisticsCollection.scala:236)
        at com.nvidia.spark.rapids.delta.GpuStatisticsCollection$.batchStatsToRow(GpuStatisticsCollection.scala:216)
        at com.databricks.sql.transaction.tahoe.rapids.GpuOptimisticTransaction.$anonfun$getOptionalGpuStatsTrackerAndStatsCollection$1(GpuOptimisticTransaction.scala:122)
        at com.databricks.sql.transaction.tahoe.rapids.GpuOptimisticTransaction.$anonfun$getOptionalGpuStatsTrackerAndStatsCollection$1$adapted(GpuOptimisticTransaction.scala:121)
        at com.nvidia.spark.rapids.delta.GpuDeltaTaskStatisticsTracker.newBatch(GpuDeltaTaskStatisticsTracker.scala:164)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.$anonfun$write$20(GpuFileFormatDataWriter.scala:553)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.$anonfun$write$20$adapted(GpuFileFormatDataWriter.scala:553)
        at scala.collection.immutable.List.foreach(List.scala:431)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.$anonfun$write$19(GpuFileFormatDataWriter.scala:553)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.$anonfun$write$19$adapted(GpuFileFormatDataWriter.scala:552)
        at com.nvidia.spark.rapids.Arm.closeOnExcept(Arm.scala:87)
        at com.nvidia.spark.rapids.Arm.closeOnExcept$(Arm.scala:85)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.closeOnExcept(GpuFileFormatDataWriter.scala:217)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.$anonfun$write$13(GpuFileFormatDataWriter.scala:552)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.$anonfun$write$13$adapted(GpuFileFormatDataWriter.scala:479)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.write(GpuFileFormatDataWriter.scala:479)
        at org.apache.spark.sql.rapids.GpuDynamicPartitionDataSingleWriter.write(GpuFileFormatDataWriter.scala:425)
        at org.apache.spark.sql.rapids.GpuFileFormatDataWriter.writeWithIterator(GpuFileFormatDataWriter.scala:86)
        at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$executeTask$1(GpuFileFormatWriter.scala:336)
        at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1731)
        at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:343)
        ... 25 more

```

--- a/delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaIdentityColumnStatsTracker.scala
+++ b/delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaIdentityColumnStatsTracker.scala
@@ -19,8 +19,8 @@ package com.nvidia.spark.rapids.delta
 import scala.collection.mutable
 
 import ai.rapids.cudf.ColumnView
+import com.nvidia.spark.rapids.{GpuColumnVector, GpuScalar}
 import com.nvidia.spark.rapids.Arm.withResource
-import com.nvidia.spark.rapids.GpuScalar
 import com.nvidia.spark.rapids.delta.shims.ShimJsonUtils
 
 import org.apache.spark.sql.catalyst.InternalRow
@@ -28,6 +28,7 @@ import org.apache.spark.sql.catalyst.expressions.{Attribute, BindReferences, Bou
 import org.apache.spark.sql.catalyst.expressions.aggregate.DeclarativeAggregate
 import org.apache.spark.sql.execution.datasources.WriteTaskStats
 import org.apache.spark.sql.types.NullType
+import org.apache.spark.sql.vectorized.ColumnarBatch
 
 class GpuDeltaIdentityColumnStatsTracker(
     val dataSchema: Seq[Attribute],
@@ -61,7 +62,7 @@ object GpuDeltaIdentityColumnStatsTracker {
   def batchStatsToRow(
       dataCols: Seq[Attribute],
       identityStatsExpr: Expression,
-      identityInfo: Seq[(String, Boolean)]): (Array[ColumnView], InternalRow) => Unit = {
+      identityInfo: Seq[(String, Boolean)]): (ColumnarBatch, InternalRow) => Unit = {
     val aggregates = identityStatsExpr.collect {
       case ae: DeclarativeAggregate => ae
     }
@@ -73,7 +74,8 @@ object GpuDeltaIdentityColumnStatsTracker {
     assert(identityInfo.size == boundRefs.size,
       s"expected ${identityInfo.size} refs found ${boundRefs.size}")
     val zipped = identityInfo.map(_._2).zip(boundRefs).zipWithIndex
-    (columnViews: Array[ColumnView], row: InternalRow) => {
+    (batch: ColumnarBatch, row: InternalRow) => {
+      val columnViews = GpuColumnVector.extractBases(batch).asInstanceOf[Array[ColumnView]]
       zipped.foreach { case ((useMax, ref), i) =>
         val cview = columnViews(ref.ordinal)
         val gpuScalar = if (useMax) {

--- a/delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaTaskStatisticsTracker.scala
+++ b/delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaTaskStatisticsTracker.scala
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, NVIDIA CORPORATION.
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
  *
  * This file was derived from DataSkippingStatsTracker.scala
  * in the Delta Lake project at https://github.com/delta-io/delta.
@@ -23,8 +23,6 @@ package com.nvidia.spark.rapids.delta
 
 import scala.collection.mutable
 
-import ai.rapids.cudf.ColumnView
-import com.nvidia.spark.rapids.GpuColumnVector
 import com.nvidia.spark.rapids.delta.shims.ShimJoinedProjection
 import org.apache.hadoop.fs.Path
 
@@ -55,7 +53,7 @@ case class GpuDeltaFileStatistics(stats: Map[String, String]) extends WriteTaskS
 class GpuDeltaTaskStatisticsTracker(
     dataCols: Seq[Attribute],
     statsColExpr: Expression,
-    batchStatsToRow: (Array[ColumnView], InternalRow) => Unit)
+    batchStatsToRow: (ColumnarBatch, InternalRow) => Unit)
     extends ColumnarWriteTaskStatsTracker {
 
   protected[this] val submittedFiles = mutable.HashMap[String, InternalRow]()
@@ -158,10 +156,7 @@ class GpuDeltaTaskStatisticsTracker(
     val aggBuffer = submittedFiles(filePath)
     extendedRow.update(0, aggBuffer)
 
-    // filter out the partition columns from the batch
-    val columnViews = GpuColumnVector.extractBases(batch).asInstanceOf[Array[ColumnView]]
-
-    batchStatsToRow(columnViews, gpuResultsBuffer)
+    batchStatsToRow(batch, gpuResultsBuffer)
 
     extendedRow.update(1, gpuResultsBuffer)
     mergeStats.target(aggBuffer).apply(extendedRow)
@@ -187,7 +182,7 @@ class GpuDeltaTaskStatisticsTracker(
 class GpuDeltaJobStatisticsTracker(
     val dataCols: Seq[Attribute],
     val statsColExpr: Expression,
-    val batchStatsToRow: (Array[ColumnView], InternalRow) => Unit)
+    val batchStatsToRow: (ColumnarBatch, InternalRow) => Unit)
     extends ColumnarWriteJobStatsTracker {
 
   var recordedStats: Map[String, String] = _

--- a/delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuStatisticsCollection.scala
+++ b/delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuStatisticsCollection.scala
@@ -24,8 +24,8 @@ package com.nvidia.spark.rapids.delta
 import scala.collection.mutable.ArrayBuffer
 
 import ai.rapids.cudf.{ColumnView, DType}
+import com.nvidia.spark.rapids.{GpuColumnVector, GpuScalar}
 import com.nvidia.spark.rapids.Arm.withResource
-import com.nvidia.spark.rapids.GpuScalar
 import com.nvidia.spark.rapids.delta.shims.{ShimDeltaColumnMapping, ShimDeltaUDF, ShimUsesMetadataFields}
 
 import org.apache.spark.sql.Column
@@ -33,6 +33,7 @@ import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute
 import org.apache.spark.sql.functions.{count, lit, max, min, struct, substring, sum, when}
 import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
 
 /** GPU version of Delta Lake's StatisticsCollection. */
 trait GpuStatisticsCollection extends ShimUsesMetadataFields {
@@ -40,6 +41,10 @@ trait GpuStatisticsCollection extends ShimUsesMetadataFields {
   def dataSchema: StructType
   val numIndexedCols: Int
   val stringPrefixLength: Int
+
+  // Build a mapping of a field path to a field index within the parent struct
+  lazy val explodedDataSchema: Map[Seq[String], Int] =
+    GpuStatisticsCollection.explode(dataSchema).toMap
 
   /**
    * statCollectionSchema is the schema that is composed of all the columns that have the stats
@@ -57,6 +62,9 @@ trait GpuStatisticsCollection extends ShimUsesMetadataFields {
    * Returns a struct column that can be used to collect statistics for the current
    * schema of the table.
    * The types we keep stats on must be consistent with DataSkippingReader.SkippingEligibleLiteral.
+   * If a column is missing from dataSchema (which will be filled with nulls), we will only
+   * collect the NULL_COUNT stats for it as the number of rows.
+   *
    * @note The stats collected here must be kept in sync with how stats are gathered by the GPU
    *       in batchStatsToRow.
    */
@@ -66,26 +74,27 @@ trait GpuStatisticsCollection extends ShimUsesMetadataFields {
       count(new Column("*")) as NUM_RECORDS,
       collectStats(MIN, statCollectionSchema) {
         // Truncate string min values as necessary
-        case (c, GpuSkippingEligibleDataType(StringType)) =>
+        case (c, GpuSkippingEligibleDataType(StringType), true) =>
           substring(min(c), 0, stringPrefixLength)
 
         // Collect all numeric min values
-        case (c, GpuSkippingEligibleDataType(_)) =>
+        case (c, GpuSkippingEligibleDataType(_), true) =>
           min(c)
       },
       collectStats(MAX, statCollectionSchema) {
         // Truncate and pad string max values as necessary
-        case (c, GpuSkippingEligibleDataType(StringType)) =>
+        case (c, GpuSkippingEligibleDataType(StringType), true) =>
           val udfTruncateMax = ShimDeltaUDF.stringStringUdf(
             GpuStatisticsCollection.truncateMaxStringAgg(prefixLength)_)
           udfTruncateMax(max(c))
 
         // Collect all numeric max values
-        case (c, GpuSkippingEligibleDataType(_)) =>
+        case (c, GpuSkippingEligibleDataType(_), true) =>
           max(c)
       },
       collectStats(NULL_COUNT, statCollectionSchema) {
-        case (c, _) => sum(when(c.isNull, 1).otherwise(0))
+        case (c, _, true) => sum(when(c.isNull, 1).otherwise(0))
+        case (_, _, false) => count(new Column("*"))
       }
     ) as 'stats
   }
@@ -124,39 +133,48 @@ trait GpuStatisticsCollection extends ShimUsesMetadataFields {
    * When `function` is not defined, that column is skipped.
    *
    * @param name     The name of the top level column for this statistic (i.e. minValues).
-   * @param schema   The schema of the data to collect statistics from.
-   * @param function A partial function that is passed both a column and metadata about that
-   *                 column. Based on the metadata, it can decide if the given statistic
-   *                 should be collected by returning the correct aggregate expression.
+   * @param schema The schema of the data to collect statistics from.
+   * @param function A partial function that is passed a tuple of (column, metadata about that
+   *                 column, a flag that indicates whether the column is in the data schema). Based
+   *                 on the metadata and flag, the function can decide if the given statistic should
+   *                 be collected on the column by returning the correct aggregate expression.
    */
   private def collectStats(
       name: String,
       schema: StructType)(
-      function: PartialFunction[(Column, StructField), Column]): Column = {
+      function: PartialFunction[(Column, StructField, Boolean), Column]): Column = {
 
     def collectStats(
         schema: StructType,
         parent: Option[Column],
-        function: PartialFunction[(Column, StructField), Column]): Seq[Column] = {
+        parentFields: Seq[String],
+        function: PartialFunction[(Column, StructField, Boolean), Column]): Seq[Column] = {
       schema.flatMap {
         case f @ StructField(name, s: StructType, _, _) =>
           val column = parent.map(_.getItem(name))
               .getOrElse(new Column(UnresolvedAttribute.quoted(name)))
-          val stats = collectStats(s, Some(column), function)
+          val stats = collectStats(s, Some(column), parentFields :+ name, function)
           if (stats.nonEmpty) {
             Some(struct(stats: _*) as ShimDeltaColumnMapping.getPhysicalName(f))
           } else {
             None
           }
         case f @ StructField(name, _, _, _) =>
+          val fieldPath = parentFields :+ name
           val column = parent.map(_.getItem(name))
               .getOrElse(new Column(UnresolvedAttribute.quoted(name)))
-          // alias the column with its physical name
-          function.lift((column, f)).map(_.as(ShimDeltaColumnMapping.getPhysicalName(f)))
+          // Note: explodedDataSchema comes from dataSchema. In the read path, dataSchema comes
+          // from the table's metadata.dataSchema, which is the same as tableDataSchema. In the
+          // write path, dataSchema comes from the DataFrame schema. We then assume
+          // TransactionWrite.writeFiles has normalized dataSchema, and
+          // TransactionWrite.getStatsSchema has done the column mapping for tableDataSchema and
+          // dropped the partition columns for both dataSchema and tableDataSchema.
+          function.lift((column, f, explodedDataSchema.contains(fieldPath))).
+              map(_.as(ShimDeltaColumnMapping.getPhysicalName(f)))
       }
     }
 
-    val allStats = collectStats(schema, None, function)
+    val allStats = collectStats(schema, None, Nil, function)
     val stats = if (numIndexedCols > 0) {
       allStats.take(numIndexedCols)
     } else {
@@ -193,61 +211,82 @@ object GpuStatisticsCollection {
 
   def batchStatsToRow(
       schema: StructType,
-      batch: Array[ColumnView],
+      explodedDataSchema: Map[Seq[String], Int],
+      batch: ColumnarBatch,
       row: InternalRow): Unit = {
     var nextSlot = 0
     // NUM_RECORDS
-    val numRows = if (batch.nonEmpty) batch(0).getRowCount.toInt else 0
-    row.setLong(nextSlot, numRows)
+    row.setLong(nextSlot, batch.numRows)
     nextSlot += 1
 
+    val views = GpuColumnVector.extractBases(batch).asInstanceOf[Array[ColumnView]]
+
     // MIN
-    nextSlot = batchStatsToRow(schema, batch, row, nextSlot, GpuStatisticsCollection.computeMin,
-      isForAllTypes = false)
+    nextSlot = batchStatsToRow(schema, explodedDataSchema, batch.numRows, views, row,
+      Nil, nextSlot, GpuStatisticsCollection.computeMin, isForAllTypes = false)
 
     // MAX
-    nextSlot = batchStatsToRow(schema, batch, row, nextSlot, GpuStatisticsCollection.computeMax,
-      isForAllTypes = false)
+    nextSlot = batchStatsToRow(schema, explodedDataSchema, batch.numRows, views, row,
+      Nil, nextSlot, GpuStatisticsCollection.computeMax, isForAllTypes = false)
 
     // NULL_COUNT
-    nextSlot = batchStatsToRow(schema, batch, row, nextSlot,
-      GpuStatisticsCollection.computeNullCount, isForAllTypes = true)
+    nextSlot = batchStatsToRow(schema, explodedDataSchema, batch.numRows, views, row,
+      Nil, nextSlot, GpuStatisticsCollection.computeNullCount, isForAllTypes = true)
 
     require(nextSlot == row.numFields, s"Expected ${row.numFields} stats, only wrote $nextSlot")
   }
 
   private def batchStatsToRow(
       schema: StructType,
-      batch: Array[ColumnView],
+      explodedDataSchema: Map[Seq[String], Int],
+      batchNumRows: Int,
+      views: Array[ColumnView],
       row: InternalRow,
+      parentFields: Seq[String],
       slot: Int,
       statFunc: (ColumnView, DataType, InternalRow, Int) => Unit,
       isForAllTypes: Boolean): Int = {
-    withResource(ColumnView.makeStructView(batch: _*)) { structView =>
-      structStatsToRow(structView, schema, row, slot, statFunc, isForAllTypes)
+    withResource(ColumnView.makeStructView(views: _*)) { structView =>
+      structStatsToRow(explodedDataSchema, structView, schema, batchNumRows, row,
+        parentFields, slot, statFunc, isForAllTypes)
     }
   }
 
   private def structStatsToRow(
+      explodedDataSchema: Map[Seq[String], Int],
       struct: ColumnView,
       structSchema: StructType,
+      batchNumRows: Int,
       row: InternalRow,
+      parentFields: Seq[String],
       slot: Int,
       statFunc: (ColumnView, DataType, InternalRow, Int) => Unit,
       isForAllTypes: Boolean): Int = {
     require(struct.getType.equals(DType.STRUCT), s"Expected struct column, found ${struct.getType}")
     var nextSlot = slot
-    (0 until structSchema.length).foreach { i =>
-      structSchema.fields(i).dataType match {
+    structSchema.foreach { field =>
+      val fieldPath = parentFields :+ field.name
+      val dataFieldIndex = explodedDataSchema.get(fieldPath)
+      field.dataType match {
         case s: StructType =>
-          withResource(struct.getChildColumnView(i)) { c =>
-            nextSlot = structStatsToRow(c, s, row, nextSlot, statFunc, isForAllTypes)
+          // If the struct is not in the field index then there's no struct column view
+          // to use. Send down null as a stand-in, since we may need to generate null count
+          // metrics for struct field members that are missing.
+          withResource(dataFieldIndex.map(struct.getChildColumnView).orNull) { c =>
+            nextSlot = structStatsToRow(explodedDataSchema, c, s, batchNumRows, row,
+              parentFields :+ field.name, nextSlot, statFunc, isForAllTypes)
           }
         case dt if isForAllTypes || GpuSkippingEligibleDataType(dt) =>
-          withResource(struct.getChildColumnView(i)) { c =>
-            statFunc(c, dt, row, nextSlot)
+          if (dataFieldIndex.isDefined) {
+            withResource(struct.getChildColumnView(dataFieldIndex.get)) { c =>
+              statFunc(c, dt, row, nextSlot)
+            }
+            nextSlot += 1
+          } else if (isForAllTypes) {
+            // doing a top-level null count for a field not in the data schema
+            row.setLong(nextSlot, batchNumRows)
+            nextSlot += 1
           }
-          nextSlot += 1
         case _ =>
       }
     }
@@ -334,6 +373,53 @@ object GpuStatisticsCollection {
       row: InternalRow,
       slot: Int): Unit = {
     row.setLong(slot, c.getNullCount)
+  }
+
+  /**
+   * Similar to Delta Lake's SchemaMergingUtils.explode except instead of mapping a field
+   * path to a field it maps a field paths to a field ordinal.
+   *
+   * Returns pairs of (full column name path, field index) in this schema as a list.
+   * For example, a schema like:
+   * <field a>          | - a
+   * <field 1>          | | - 1
+   * <field 2>          | | - 2
+   * <field b>          | - b
+   * <field c>          | - c
+   * <field `foo.bar`>  | | - `foo.bar`
+   * <field 3>          |   | - 3
+   * will return [
+   * ([a], 0), ([a, 1], 0), ([a, 2], 1), ([b], 1),
+   * ([c], 2), ([c, foo.bar], 0), ([c, foo.bar, 3], 0)
+   * ]
+   */
+  def explode(schema: StructType): Seq[(Seq[String], Int)] = {
+    def recurseIntoComplexTypes(complexType: DataType): Seq[(Seq[String], Int)] = {
+      complexType match {
+        case s: StructType => explode(s)
+        case a: ArrayType => recurseIntoComplexTypes(a.elementType)
+            .map { case (path, i) => (Seq("element") ++ path, i) }
+        case m: MapType =>
+          recurseIntoComplexTypes(m.keyType)
+              .map { case (path, i) => (Seq("key") ++ path, i) } ++
+              recurseIntoComplexTypes(m.valueType)
+                  .map { case (path, i) => (Seq("value") ++ path, i) }
+        case _ => Nil
+      }
+    }
+
+    schema.zipWithIndex.flatMap {
+      case (StructField(name, s: StructType, _, _), i) =>
+        Seq((Seq(name), i)) ++
+            explode(s).map { case (path, i) => (Seq(name) ++ path, i) }
+      case (StructField(name, a: ArrayType, _, _), i) =>
+        Seq((Seq(name), i)) ++
+            recurseIntoComplexTypes(a).map { case (path, i) => (Seq(name) ++ path, i) }
+      case (StructField(name, m: MapType, _, _), i) =>
+        Seq((Seq(name), i)) ++
+            recurseIntoComplexTypes(m).map { case (path, i) => (Seq(name) ++ path, i) }
+      case (f, i) => (Seq(f.name), i) :: Nil
+    }
   }
 }
 

--- a/delta-lake/delta-20x/src/main/scala/org/apache/spark/sql/delta/rapids/delta20x/GpuOptimisticTransaction.scala
+++ b/delta-lake/delta-20x/src/main/scala/org/apache/spark/sql/delta/rapids/delta20x/GpuOptimisticTransaction.scala
@@ -25,7 +25,6 @@ import java.net.URI
 
 import scala.collection.mutable.ListBuffer
 
-import ai.rapids.cudf.ColumnView
 import com.nvidia.spark.rapids._
 import com.nvidia.spark.rapids.delta._
 import org.apache.commons.lang3.exception.ExceptionUtils
@@ -47,6 +46,7 @@ import org.apache.spark.sql.execution.datasources.{BasicWriteJobStatsTracker, Fi
 import org.apache.spark.sql.functions.to_json
 import org.apache.spark.sql.rapids.{BasicColumnarWriteJobStatsTracker, ColumnarWriteJobStatsTracker, GpuFileFormatWriter, GpuWriteJobStatsTracker}
 import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
 import org.apache.spark.util.{Clock, SerializableConfiguration}
 
 /**
@@ -133,8 +133,9 @@ class GpuOptimisticTransaction
         }
 
         val statsSchema = statsCollection.statCollectionSchema
-        val batchStatsToRow = (columnViews: Array[ColumnView], row: InternalRow) => {
-          GpuStatisticsCollection.batchStatsToRow(statsSchema, columnViews, row)
+        val explodedDataSchema = statsCollection.explodedDataSchema
+        val batchStatsToRow = (batch: ColumnarBatch, row: InternalRow) => {
+          GpuStatisticsCollection.batchStatsToRow(statsSchema, explodedDataSchema, batch, row)
         }
         Some(new GpuDeltaJobStatisticsTracker(statsDataSchema, statsColExpr, batchStatsToRow))
       } else {

--- a/delta-lake/delta-21x/src/main/scala/org/apache/spark/sql/delta/rapids/delta21x/GpuOptimisticTransaction.scala
+++ b/delta-lake/delta-21x/src/main/scala/org/apache/spark/sql/delta/rapids/delta21x/GpuOptimisticTransaction.scala
@@ -25,7 +25,6 @@ import java.net.URI
 
 import scala.collection.mutable.ListBuffer
 
-import ai.rapids.cudf.ColumnView
 import com.nvidia.spark.rapids._
 import com.nvidia.spark.rapids.delta._
 import org.apache.commons.lang3.exception.ExceptionUtils
@@ -47,6 +46,7 @@ import org.apache.spark.sql.execution.datasources.{BasicWriteJobStatsTracker, Fi
 import org.apache.spark.sql.functions.to_json
 import org.apache.spark.sql.rapids.{BasicColumnarWriteJobStatsTracker, ColumnarWriteJobStatsTracker, GpuFileFormatWriter, GpuWriteJobStatsTracker}
 import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
 import org.apache.spark.util.{Clock, SerializableConfiguration}
 
 /**
@@ -133,8 +133,9 @@ class GpuOptimisticTransaction
         }
 
         val statsSchema = statsCollection.statCollectionSchema
-        val batchStatsToRow = (columnViews: Array[ColumnView], row: InternalRow) => {
-          GpuStatisticsCollection.batchStatsToRow(statsSchema, columnViews, row)
+        val explodedDataSchema = statsCollection.explodedDataSchema
+        val batchStatsToRow = (batch: ColumnarBatch, row: InternalRow) => {
+          GpuStatisticsCollection.batchStatsToRow(statsSchema, explodedDataSchema, batch, row)
         }
         Some(new GpuDeltaJobStatisticsTracker(statsDataSchema, statsColExpr, batchStatsToRow))
       } else {

--- a/delta-lake/delta-22x/src/main/scala/org/apache/spark/sql/delta/rapids/delta22x/GpuOptimisticTransaction.scala
+++ b/delta-lake/delta-22x/src/main/scala/org/apache/spark/sql/delta/rapids/delta22x/GpuOptimisticTransaction.scala
@@ -25,7 +25,6 @@ import java.net.URI
 
 import scala.collection.mutable.ListBuffer
 
-import ai.rapids.cudf.ColumnView
 import com.nvidia.spark.rapids._
 import com.nvidia.spark.rapids.delta._
 import org.apache.commons.lang3.exception.ExceptionUtils
@@ -47,6 +46,7 @@ import org.apache.spark.sql.execution.datasources.{BasicWriteJobStatsTracker, Fi
 import org.apache.spark.sql.functions.to_json
 import org.apache.spark.sql.rapids.{BasicColumnarWriteJobStatsTracker, ColumnarWriteJobStatsTracker, GpuFileFormatWriter, GpuWriteJobStatsTracker}
 import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
 import org.apache.spark.util.{Clock, SerializableConfiguration}
 
 /**
@@ -117,8 +117,9 @@ class GpuOptimisticTransaction
       val statsColExpr = getGpuStatsColExpr(statsDataSchema, statsCollection)
 
       val statsSchema = statsCollection.statCollectionSchema
-      val batchStatsToRow = (columnViews: Array[ColumnView], row: InternalRow) => {
-        GpuStatisticsCollection.batchStatsToRow(statsSchema, columnViews, row)
+      val explodedDataSchema = statsCollection.explodedDataSchema
+      val batchStatsToRow = (batch: ColumnarBatch, row: InternalRow) => {
+        GpuStatisticsCollection.batchStatsToRow(statsSchema, explodedDataSchema, batch, row)
       }
       (Some(new GpuDeltaJobStatisticsTracker(statsDataSchema, statsColExpr, batchStatsToRow)),
           Some(statsCollection))

--- a/delta-lake/delta-spark321db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransaction.scala
+++ b/delta-lake/delta-spark321db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransaction.scala
@@ -25,7 +25,6 @@ import java.net.URI
 
 import scala.collection.mutable.ListBuffer
 
-import ai.rapids.cudf.ColumnView
 import com.databricks.sql.transaction.tahoe._
 import com.databricks.sql.transaction.tahoe.actions.FileAction
 import com.databricks.sql.transaction.tahoe.commands.cdc.CDCReader
@@ -48,6 +47,7 @@ import org.apache.spark.sql.functions.{col, to_json}
 import org.apache.spark.sql.rapids.{BasicColumnarWriteJobStatsTracker, ColumnarWriteJobStatsTracker, GpuFileFormatWriter, GpuWriteJobStatsTracker}
 import org.apache.spark.sql.rapids.delta.GpuIdentityColumn
 import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
 import org.apache.spark.util.{Clock, SerializableConfiguration}
 
 
@@ -147,8 +147,9 @@ class GpuOptimisticTransaction(
       }
 
       val statsSchema = statsCollection.statCollectionSchema
-      val batchStatsToRow = (columnViews: Array[ColumnView], row: InternalRow) => {
-        GpuStatisticsCollection.batchStatsToRow(statsSchema, columnViews, row)
+      val explodedDataSchema = statsCollection.explodedDataSchema
+      val batchStatsToRow = (batch: ColumnarBatch, row: InternalRow) => {
+        GpuStatisticsCollection.batchStatsToRow(statsSchema, explodedDataSchema, batch, row)
       }
       Some(new GpuDeltaJobStatisticsTracker(statsDataSchema, statsColExpr, batchStatsToRow))
     } else {

--- a/delta-lake/delta-spark330db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransaction.scala
+++ b/delta-lake/delta-spark330db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransaction.scala
@@ -25,7 +25,6 @@ import java.net.URI
 
 import scala.collection.mutable.ListBuffer
 
-import ai.rapids.cudf.ColumnView
 import com.databricks.sql.transaction.tahoe._
 import com.databricks.sql.transaction.tahoe.actions.{AddFile, FileAction}
 import com.databricks.sql.transaction.tahoe.constraints.{Constraint, Constraints}
@@ -47,6 +46,7 @@ import org.apache.spark.sql.functions.to_json
 import org.apache.spark.sql.rapids.{BasicColumnarWriteJobStatsTracker, ColumnarWriteJobStatsTracker, GpuFileFormatWriter, GpuWriteJobStatsTracker}
 import org.apache.spark.sql.rapids.delta.GpuIdentityColumn
 import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
 import org.apache.spark.util.{Clock, SerializableConfiguration}
 
 /**
@@ -118,8 +118,9 @@ class GpuOptimisticTransaction(
       val statsColExpr = getGpuStatsColExpr(statsDataSchema, statsCollection)
 
       val statsSchema = statsCollection.statCollectionSchema
-      val batchStatsToRow = (columnViews: Array[ColumnView], row: InternalRow) => {
-        GpuStatisticsCollection.batchStatsToRow(statsSchema, columnViews, row)
+      val explodedDataSchema = statsCollection.explodedDataSchema
+      val batchStatsToRow = (batch: ColumnarBatch, row: InternalRow) => {
+        GpuStatisticsCollection.batchStatsToRow(statsSchema, explodedDataSchema, batch, row)
       }
       (Some(new GpuDeltaJobStatisticsTracker(statsDataSchema, statsColExpr, batchStatsToRow)),
           Some(statsCollection))

--- a/integration_tests/src/main/python/delta_lake_write_test.py
+++ b/integration_tests/src/main/python/delta_lake_write_test.py
@@ -887,3 +887,38 @@ def test_delta_write_optimized_partitioned(spark_tmp_path):
     do_write(confs)
     opmetrics = get_last_operation_metrics(gpu_data_path)
     assert int(opmetrics["numFiles"]) == 2
+
+@allow_non_gpu(*delta_meta_allow)
+@delta_lake
+@ignore_order
+@pytest.mark.skipif(is_before_spark_320(), reason="Delta Lake writes are not supported before Spark 3.2.x")
+def test_delta_write_partial_overwrite_replace_where(spark_tmp_path):
+    gen_list = [("a", int_gen),
+                ("b", SetValuesGen(StringType(), ["x", "y", "z"])),
+                ("c", string_gen),
+                ("d", SetValuesGen(IntegerType(), [1, 2, 3])),
+                ("e", long_gen)]
+    data_path = spark_tmp_path + "/DELTA_DATA"
+    assert_gpu_and_cpu_writes_are_equal_collect(
+        lambda spark, path: gen_df(spark, gen_list).coalesce(1).write.format("delta")\
+            .partitionBy("b", "d")\
+            .save(path),
+        lambda spark, path: spark.read.format("delta").load(path),
+        data_path,
+        conf=copy_and_update(writer_confs, delta_writes_enabled_conf))
+    with_cpu_session(lambda spark: assert_gpu_and_cpu_delta_logs_equivalent(spark, data_path))
+    # overwrite with a subset of the original schema
+    gen_list = [("b", SetValuesGen(StringType(), ["y"])),
+                ("e", long_gen),
+                ("c", string_gen),
+                ("d", SetValuesGen(IntegerType(), [1, 2, 3]))]
+    assert_gpu_and_cpu_writes_are_equal_collect(
+        lambda spark, path: gen_df(spark, gen_list).coalesce(1).write.format("delta")\
+            .mode("overwrite")\
+            .partitionBy("b", "d")\
+            .option("replaceWhere", "b = 'y'")\
+            .save(path),
+        lambda spark, path: spark.read.format("delta").load(path),
+        data_path,
+        conf=copy_and_update(writer_confs, delta_writes_enabled_conf))
+    with_cpu_session(lambda spark: assert_gpu_and_cpu_delta_logs_equivalent(spark, data_path))

